src/hotspot/share/opto/loopnode.cpp

```diff
@@ -4695,6 +4708,329 @@ void PhaseIdealLoop::build_and_optimize() {
   }
 }
 
+// Expansion of ScopedValue nodes happen during loop opts because their expansion creates an opportunity for
+// further loop optimizations (see comment in LateInlineScopedValueCallGenerator::process_result)
+bool PhaseIdealLoop::expand_scoped_value_get_nodes() {
+  bool progress = false;
+  assert(!_igvn.delay_transform(), "");
+  _igvn.set_delay_transform(true);
+  for (uint i = _scoped_value_get_nodes.size(); i > 0; i--) {
+    Node* n = _scoped_value_get_nodes.at(i - 1);
+    if (n->Opcode() == Op_ScopedValueGetResult) {
+      // Remove the ScopedValueGetResult entirely
+      ScopedValueGetResultNode* get_result = (ScopedValueGetResultNode*) n;
+      Node* result_out = get_result->result_out();
+      Node* result_in = get_result->in(ScopedValueGetResultNode::GetResult);
+      if (result_out != nullptr) {
+        _igvn.replace_node(result_out, result_in);
+      } else {
+        _igvn.replace_input_of(get_result, ScopedValueGetResultNode::GetResult, C->top());
+      }
+      lazy_replace(get_result->control_out(), get_result->in(ScopedValueGetResultNode::Control));
+      progress = true;
+      remove_scoped_value_get_at(i-1);
+    }
+  }
+  while (_scoped_value_get_nodes.size() > 0) {
+    Node* n = _scoped_value_get_nodes.pop();
+    assert (n->Opcode() == Op_ScopedValueGetHitsInCache, "");
+    ScopedValueGetHitsInCacheNode* get_from_cache = (ScopedValueGetHitsInCacheNode*) n;
+    expand_get_from_sv_cache(get_from_cache);
+    progress = true;
+  }
+  _igvn.set_delay_transform(false);
+  return progress;
+}
+
+void PhaseIdealLoop::expand_get_from_sv_cache(ScopedValueGetHitsInCacheNode* get_from_cache) {
+  get_from_cache->verify();
+#ifdef ASSERT
+  for (DUIterator_Fast imax, i = get_from_cache->fast_outs(imax); i < imax; i++) {
+    Node* u = get_from_cache->fast_out(i);
+    assert(u->is_Bool() || u->Opcode() == Op_ScopedValueGetLoadFromCache, "");
+  }
+#endif
+  BoolNode* bol = get_from_cache->find_unique_out_with(Op_Bool)->as_Bool();
+  assert(bol->_test._test == BoolTest::ne, "");
+  IfNode* iff = bol->find_unique_out_with(Op_If)->as_If();
+  ProjNode* success = iff->proj_out(1);
+  ProjNode* failure = iff->proj_out(0);
+
+
+  ScopedValueGetLoadFromCacheNode* load_from_cache = (ScopedValueGetLoadFromCacheNode*)success->find_unique_out_with(Op_ScopedValueGetLoadFromCache);
+  if (load_from_cache != nullptr) {
+    load_from_cache->verify();
+  }
+  Node* first_index = get_from_cache->index1();
+  Node* second_index = get_from_cache->index2();
+
+  if (first_index == C->top() && second_index == C->top()) {
```

Comment:
could this not be done during igvn?

Reply:
What happens here is that the cache was always seen to be null so no code to to probe the cache was added to the IR. When that happens, the optimizations still apply (i.e. there could be dominated `ScopedValue.get()` that can be replaced by this one or a dominating one that can replace this one). This should also be very uncommon. So the nodes that are put in place to enable optimizations should be left in until expansion to not miss optimization opportunities.